doc/developer/design/20320328_persist_columnar.md

```diff
@@ -0,0 +1,917 @@
+```
+
+To migrate to the new structured data I propose we extend the above schema to:
+
+```
+{
+  "k": bytes,
+  "v": bytes,
+  "t": i64,
+  "d": i64,
+  "k_s": {
+    // ... nested structured data
+  },
+  "v_s": {
+    // ... nested structured data
+  },
+}
+```
+
+In other words, we dual encode blobs with both `Codec` and Arrow and write them
+to a single Parquet file. This has the following benefits:
+
+1. Slow rollout/shadow migration. If we write both formats we can let this
+   change bake for as long as we would like, validating that `Codec` and Arrow
+   decode to the same `Row`s. Reading only Arrow data could be turned on or off
+   via a LaunchDarkly flag.
+2. Incremental progress. If we write both formats then we could read back Arrow
+   data but sort and consolidate with the `Codec` format. Currently decoding
+   protobuf (i.e. `ProtoDatum`) is computationally expensive, getting an
+   incremental win of improving decode time would be nice.
+3. Doesn't increase the number of `PUT` or `GET` requests to `S3`.
+4. Doesn't require any change to state stored in Consensus.
+5. (Forward looking) When we switch to reading only Arrow data we can leverage
+   Parquet's projection pushdown to ignore the old `'k'` and `'v'` columns.
+
+The current max blob size is 128MiB, we have two options for how we want to
+handle the "max" size when we write both `Codec` data and Arrow at the same
+time.
+
+1. Ignore Arrow when calculating the max. This increases the size of S3 objects
+   which may impact performance for things that have been tuned to assume
+   128MiB sized objets, e.g. compute back pressure.
+2. Include Arrow when calculating the max. This reduces the amount of rows that
+   can get stored in a single blob, thus increasing the number of Parts
+   per-Run. This increases the number of S3 `GET`s and `PUT`s which could
+   impact cost, and requires `ConsolidatingIter` to handle a larger number of
+   Parts.
+
+Neither option is great, but I am leanning towards [1] since it puts the
+pressure/complexity on S3 and the network more than Persist internals.
+
+### Consolidation and Compaction
+
+To start we can write both the `Codec` and Arrow versions of the data, and do
+compaction based on the bytes from `Codec`. Eventually we'll need to
+consolidate the Arrow data directly which requires sorting the data. Sorting
+an Arrow `StructArray` is possible (e.g. the `pyarrow` library [supports it](https://arrow.apache.org/docs/python/generated/pyarrow.StructArray.html#pyarrow.StructArray.sort))
+but neither the `arrow-rs` or `arrow2` crates support it. I filed an
+[issue with `arrow-rs`](https://github.com/apache/arrow-rs/issues/5559) but
+realistically this is something we'll probably need to implement. Once data is
+sorted consolidating an Arrow array is relatively easy and fast! To delete a
+value all we need to do is unset the bit in the
+[validity bitmap](https://arrow.apache.org/docs/format/Columnar.html#validity-bitmaps)
+for that value.
```

Comment:
Could we say a little bit about streaming consolidation -- would we support it, and if so how would it work?

Reply:
Need to think on this some bit, will get back to you!